### PR TITLE
porting tensorflow eigen threadpool scheduling

### PR DESCRIPTION
**Pull request by linlifan**
# Description

oneDNN with eigen threadpool is the default mode in tensorflow. two issues have been identified which impacts on oneDNN performance with eigen threadpool.
1. scheduling from master thread
 https://github.com/tensorflow/tensorflow/pull/58071
2. recursive scheduling
https://github.com/tensorflow/tensorflow/pull/60346 

## Performance improvements

eigen before:
 DNNL_VERBOSE=0 numactl -C 0-7 ./tests/benchdnn/benchdnn --ip --mode=P --fix-times-per-prb=500  mb128ic768oc2304
Output template: perf,%engine%,%impl%,%name%,%prb%,%Gops%,%-time%,%-Gflops%,%0time%,%0Gflops%
perf,cpu,brgemm:avx512_core,,--mode=P --fix-times-per-prb=500 --ip mb128ic768oc2304,0.452985,0.396973,1141.1,0.636601,711.568
tests:1 passed:1 skipped:0 mistrusted:0 unimplemented:0 invalid_arguments:0 failed:0 listed:0
total perf: min(ms):0.396973 avg(ms):0.636601

eigen after:
DNNL_VERBOSE=0 numactl -C 0-7 ./tests/benchdnn/benchdnn --ip --mode=P --fix-times-per-prb=500  mb128ic768oc2304
Output template: perf,%engine%,%impl%,%name%,%prb%,%Gops%,%-time%,%-Gflops%,%0time%,%0Gflops%
perf,cpu,brgemm:avx512_core,,--mode=P --fix-times-per-prb=500 --ip mb128ic768oc2304,0.452985,0.390869,1158.92,0.416582,1087.39
tests:1 passed:1 skipped:0 mistrusted:0 unimplemented:0 invalid_arguments:0 failed:0 listed:0
total perf: min(ms):0.390869 avg(ms):0.416582

omp:
DNNL_VERBOSE=0 numactl -C 0-7 ./tests/benchdnn/benchdnn --ip --mode=P --fix-times-per-prb=500  mb128ic768oc2304
Output template: perf,%engine%,%impl%,%name%,%prb%,%Gops%,%-time%,%-Gflops%,%0time%,%0Gflops%
perf,cpu,brgemm:avx512_core,,--mode=P --fix-times-per-prb=500 --ip mb128ic768oc2304,0.452985,0.360352,1257.06,0.36512,1240.65
tests:1 passed:1 skipped:0 mistrusted:0 unimplemented:0 invalid_arguments:0 failed:0 listed:0
total perf: min(ms):0.360352 avg(ms):0.36512

### CODE REVIEW

**Comment by vpirogov** [comment]
@linlifan, the changes in this PR apply to threadpool implementation used in oneDNN tests only. These will not have any effect on Tensorflow performance. Could you please elaborate on why you want to change the test?

**Comment by linlifan** [comment]
@vpirogov It confused people who is running oneDNN with eigen threadpool. the performance gap topic is flying all the time. and it may also impact oneDNN graph performance.  questions mights comes up time to time. I suggest to have a consistent benchmark setup to deliever oneDNN performance.

**Comment by vpirogov** [comment]
This is test code, so it will not change oneDNN Graph behavior either. The only place where these changes make difference is benchdnn performance results. 

Could you please share performance data that shows that benchdnn results with your changes are consistent with performance observed in Tensorflow?

**Comment by linlifan** [comment]
@vpirogov agree that this is test code. but it will be more convenient for develop to benchmark performance. please find the tf2.12 pip result.
image

File: CMakeLists.txt [hunk 1]
```diff
@@ -139,7 +139,7 @@ if(UNIX OR MINGW)
         set(CMAKE_C_FLAGS "-std=c99 ${CMAKE_C_FLAGS}")
         # Let SYCL to choose the C++ standard it needs.
         if(NOT DNNL_WITH_SYCL)
-            set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
+	set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
```
Comment (vpirogov) [inline CMakeLists.txt:142]: Why changing formatting here?

File: CMakeLists.txt [hunk 2]
```diff
@@ -148,7 +148,7 @@ if(UNIX OR MINGW)
             set(CMAKE_C_STANDARD 99)
         endif()
         if(NOT DEFINED CMAKE_CXX_STANDARD AND NOT DNNL_WITH_SYCL)
-            set(CMAKE_CXX_STANDARD 11)
+            set(CMAKE_CXX_STANDARD 17)
```
Comment (vpirogov) [inline CMakeLists.txt:151]: Default C++ standard for oneDNN is C++ 11. Please revert this change.

File: tests/test_thread.cpp [hunk 3]
```diff
@@ -144,6 +200,58 @@ class threadpool_t : public dnnl::threadpool_interop::threadpool_iface {
                     fn(j, n);
             });
         }
+*/
```
Comment (mgouicem) [inline tests/test_thread.cpp:203]: please don't leave commented code.

File: tests/test_thread.cpp [hunk 4]
```diff
@@ -144,6 +200,58 @@ class threadpool_t : public dnnl::threadpool_interop::threadpool_iface {
...
+#if 0
+    for (int i = 0; i < njobs_to_schedule; i++) {
+          tp_->ScheduleWithHint(
+          [balance, i, n, njobs, fn]() { run_jobs(balance, i, n, njobs, fn); },
+          i, i + 1);
+    }
+    if (use_caller_thread) {
+      run_jobs(balance, njobs - 1, n, njobs, fn);
+    }
+#else
```
Comment (mgouicem) [inline tests/test_thread.cpp:231]: any reason to keep this dead code?

**Review by mgouicem — COMMENTED** [review]
Thanks for the contribution.
Could you please:
- squash both commits and align commit message to [oneDNN guideline](https://github.com/oneapi-src/oneDNN/blob/master/CONTRIBUTING.md#code-contribution-guidelines).
- run clang-format on your patch.

File: tests/test_thread.cpp (hunk 4 above)
Comment (linlifan) [reply tests/test_thread.cpp]: addressed , thank you

**Comment by vpirogov** [comment]
Closing as stale.
